### PR TITLE
Refactoring

### PR DESCRIPTION
**Pull request by stewart**
This PR substancially re-writes cppp-io.js, offering a simpler abstraction for protocols, and a unified interface for using them.

Example usage:

``` javascript
var Threepio = require("./");

var client = new Threepio("http", { host: "localhost", port: "3000" });
// var client = new Threepio("socket-io", { host: "localhost", port: "4000" });
// var client = new Threepio("mqtt", { broker: "mqtt://localhost });

client.connect(function(err, mcp) {
  if (err) { throw err; }

  console.log(mcp);

  mcp.echo(["Hello, world!"], console.log);
  // => null, "Hello, world!"

  mcp.robot("TestBot").on("hello", function() { console.log("hello event"); });
  mcp.robot("TestBot").hello(["dave"], console.log);
  // => null, "Hello, Dave!"

  mcp.robot("TestBot").device("ping").on("ping", console.log);
  mcp.robot("TestBot").device("ping").ping();
});
```

New system components:
#### Client

The Client is a thin wrapper around the Protocols. It is the primary user-facing component of cppp-io.js.

It checks for invalid protocols on startup, and ensures the host/port options are provided.

It also supplies callback-taking methods that proxy to the used Protocol's `#connect`, `#update`, and `#disconnect` methods.

It fetches data from the Protocol, and decorates it using the Decorator.
#### Protocols

A Protocol implementation obeys the following spec implementation, and contains implementation details for connecting with a specific API backend (e.g. HTTP, Socket.IO, MQTT):

``` javascript
/**
 * Creates a new Protocol instance
 *
 * @constructor Protocol
 * @param {Object} opts - protocol options
 */
var Protocol = module.exports = function Protocol(opts) {
  this.connected = false;
  this.data = null;

  this.host = opts.host;
  this.port = opts.port;

  if (!this.host) {
    throw new Error("No host supplied for protocol");
  }

  if (!this.port) {
    throw new Error("No port supplied for protocol");
  }
};

/**
 * Performs first-time setup / data fetching for the Protocol.
 *
 * Callback is invoked when complete, with the arguments (err, data)
 *
 * @param {Function} callback - function to invoke when connected + data fetched
 * @return {void}
 */
Protocol.prototype.connect = function(callback) {
};

/**
 * Performs teardown, disconnects from remote.
 *
 * Callback is invoked when complete, with an error if any occurred.
 *
 * @param {Function} callback - function to invoke when disconnected.
 * @return {void}
 */
Protocol.prototype.disconnect = function(callback) {
};

/**
 * Fetches new data from the server, replacing existing data entirely.
 *
 * Callback is invoked when complete, with the arguments (err, data)
 *
 * @param {Function} callback - function to invoke when updated
 * @return {void}
 */
Protocol.prototype.update = function(callback) {
};

/**
 * Executes a command on the server at the specified endpoint.
 *
 * @param {String} name - name of the command to run
 * @param {String} endpoint - CPPP-IO endpoint path the command should run at
 * @param {Object} [params] - command parameters
 * @param {Function} [callback] - function to invoke with command result
 * @return {void}
 */
Protocol.prototype.command = function(name, endpoint, params, callback) {
};

/**
 * Listens for an event coming from the server, triggering a supplied callback
 * every time the event occurs, passing along any data the server emits.
 *
 * @param {String} name - name of the event to listen for
 * @param {String} endpoint - CPPP-IO endpoint path the event exists at
 * @param {Function} callback - function to invoke when event is received
 * @return {void}
 */
Protocol.prototype.event = function(name, endpoint, callback) {
};
```

For a more detailed example, see [the HTTP protocol implementation](https://github.com/hybridgroup/cppp-io.js/blob/4314c66ba1f18d562b4daf53f6d97397300e07b5/lib/protocols/http.js)
#### Decorator

The Decorator provides a function, `#decorate`.

This function takes as an argument the protocol instance being used, and a copy of the data the protocol has yielded.

It then decorates this copy with useful accessors and shortcut methods, e.g. an `#on` method for listening to MCP/Robot/Device events, and a `#command` method to easily invoke commands on the same.

It also adds utility methods to the provided MCP and each robot, `mcp#robot(name)` and `robot#device(string)`. These provide easier ways to navigate the MCP tree than manually searching through arrays for the device/robot you want.
## TODO

Not part of this PR, but these are the next steps I believe would be nice to focus on:
- MQTT support
- easier command syntax (`robot.command("arg", "arg2", 10, function(data) {})`)

### CODE REVIEW

File: lib/protocols/socket-io.js [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+    throw new Error("No host supplied for SocketIO protocol");
+  }
+
+  if (!this.port) {
+    throw new Error("No port supplied for SocketIO protocol");
+  }
+
+  Object.defineProperties(this, {
+    url: property(function() {
+      return "http://" + this.host + ":" + this.port;
```
Comment (luisgodinez) [inline lib/protocols/socket-io.js:36]: @stewart do we need "http://" here? i don't think we needed for socket-io.

Comment (stewart) [inline lib/protocols/socket-io.js:36]: WebSocket connections upgrade from plain HTTP, and Socket.IO encourages using `http` or `https` to make fallbacks easier. I was also following [Edgar's examples](https://github.com/hybridgroup/cylon-api-socketio/blob/master/examples/generic_message_event/simple-device-client.html#L29). At some point I'd also like to see if we can enable auth or https for one or more of these protocols.

**Comment by deadprogram** [comment]
This is pretty incredible stuff. Let's discuss so we can merge.

**Comment by stewart** [comment]
MQTT support waiting in branch on possible patch to API plugin, will then merge it into here - SocketIO and HTTP support is good, although I think both SocketIO and MQTT plugins will need a release before this will work with them.

**Comment by stewart** [comment]
MQTT support is in - release of API plugin updates should still precede this, but other than that I think we're good to go.

**Pull request merged**